### DISCUSSION TITLE
Android Gradle plugin requires Java 17 to run. You are currently using Java 11.

### DISCUSSION THREAD

**Discussion by gerken-tss** [codemagic-ci-cd/codemagic-docs#2336] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

I have a project in which I recently updated my Gradle dependencies as follows (`android/build.gradle`):
```
buildscript {
    ext.kotlin_version = '1.9.22'

   … // some stuff that is not relevant for this issue

    dependencies {
        classpath 'com.android.tools.build:gradle:8.2.2'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    }
}
```

Content of `android/gradle/wrapper/gradle-wrapper.properties`: 

```
…
distributionUrl=https\://services.gradle.org/distributions/gradle-8.5-all.zip
```

Locally, I can build the problem without any issues, using the Java version shipped with Android Studio. In the Codemagic pipeline, it leads to an error saying that Java 11 is being used whereas Java 17 is needed.

If I set the `java` environment key to `1.8` like it's described in the [docs](https://docs.codemagic.io/yaml-basic-configuration/yaml-getting-started/), it fails with a different error ("Could not resolve com.android.tools.build:gradle:8.2.2" -> see "Actual results")

### Expected results

No errors during build

### Actual results

```
* Where:
Build file '/Users/builder/clone/android/app/build.gradle' line: 24

* What went wrong:
A problem occurred evaluating project ':app'.
> Failed to apply plugin 'com.android.internal.application'.
   > Android Gradle plugin requires Java 17 to run. You are currently using Java 11.
      Your current JDK is located in /Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home
      You can try some of the following options:
       - changing the IDE settings.
       - changing the JAVA_HOME environment variable.
       - changing `org.gradle.java.home` in `gradle.properties`.
```

After setting `java` to `1.8` in the `codemagic.yaml`:

```
FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'android'.
> Could not resolve all files for configuration ':classpath'.
   > Could not resolve com.android.tools.build:gradle:8.2.2.
     Required by:
         project :
      > No matching variant of com.android.tools.build:gradle:8.2.2 was found. The consumer was configured to find a library for use during runtime, compatible with Java 8, packaged as a jar, and its dependencies declared externally, as well as attribute 'org.gradle.plugin.api-version' with value '8.5' but:
          - Variant 'apiElements' capability com.android.tools.build:gradle:8.2.2 declares a library, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component for use during compile-time, compatible with Java 11 and the consumer needed a component for use during runtime, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.5')
          - Variant 'javadocElements' capability com.android.tools.build:gradle:8.2.2 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.5')
          - Variant 'runtimeElements' capability com.android.tools.build:gradle:8.2.2 declares a library for use during runtime, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 11 and the consumer needed a component, compatible with Java 8
              - Other compatible attribute:
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.5')
          - Variant 'sourcesElements' capability com.android.tools.build:gradle:8.2.2 declares a component for use during runtime, and its dependencies declared externally:
              - Incompatible because this component declares documentation and the consumer needed a library
              - Other compatible attributes:
                  - Doesn't say anything about its target Java version (required compatibility with Java 8)
                  - Doesn't say anything about its elements (required them packaged as a jar)
                  - Doesn't say anything about org.gradle.plugin.api-version (required '8.5')
```

### Build id (optional)

65b2557cca74d6eba4de1b06

**Comment by gerken-tss** [comment, marked as answer]
Very weird!

Like I said, when setting the environment variable to `1.8`, there was still an error.

However, I was confused by the error that stated
> The consumer was configured to find a library for use during runtime, compatible with Java **8**

Why does it say Java 8?

I had a look at the [Codemagic docs](https://docs.codemagic.io/yaml-basic-configuration/yaml-getting-started/) again and saw the example once again:

```
environment:
  flutter: stable   # Define the channel name, version (e.g. v1.13.4), or fvm for Flutter Version Management
  xcode: latest     # Define latest, edge or version (e.g. 11.2)
  cocoapods: 1.9.1  # Define default or version
  node: 12.14.0     # Define default, latest, current, lts, carbon (or another stream), nightly or version
  npm: 6.13.7       # Define default, latest, next, lts or version
  ndk: r21d         # Define default or revision (e.g. r19c)
  java: 1.8         # Define default, or platform version (e.g. 11)
  ruby: 2.7.2       # Define default or version
```

It was weird to me that it had `1.8` (internal Java version) as the value but in the commented suggested version `11` (external Java version). This made me try to set it to `18` instead of `1.8` in the `Codemagic.yaml`.

This led to a build fail once again:

```
== Use Java version 18 ==

Build failed :|
Unable to set JAVA_HOME for Java 18, check our documentation for installed versions
```

I thought maybe this Java version is not even installed on the build machines so I tried with `java: 17`.

Et voilá - it works!

```
== Use Java version 17 ==
> java -version
openjdk version "17.0.9" 2023-10-17 LTS
OpenJDK Runtime Environment Zulu17.46+19-CA (build 17.0.9+8-LTS)
OpenJDK 64-Bit Server VM Zulu17.46+19-CA (build 17.0.9+8-LTS, mixed mode, sharing)
```

What I find interesting is that when I set it to `1.8`, the OpenJDK Runtime Environment is major version `8`, but when I set to `17` the major version is not `7` but `17`. I don't know if this has anything to do with the issue or if it doesn't matter if I set it to `17` or `1.7` but now it works.

**Reply by paulperez-dev** [reply]
I'm using the workflow editor but your answer guided me. I set the `JAVA_HOME` environment variable to the path of the version my project is using (see [documentation](https://docs.codemagic.io/specs/versions-macos/))

**Comment by fbardecio** [comment]
> I'm using the workflow editor but your answer guided me. I set the `JAVA_HOME` environment variable to the path of the version my project is using (see [documentation](https://docs.codemagic.io/specs/versions-macos/))

I'm also using the workflow editor and this worked as a charm, thank you both!
